Add fromXContent method to HighlightBuilder

cbuescher opened the pull request:
For the search refactoring the HighlightBuilder needs a way to create new instances by parsing xContent. For bwc this PR start by moving over and slightly modifying the parsing from HighlighterParseElement and keeps parsing for top level highlighter and field options separate. Also adding tests for roundtrip of random builder (rendering it to xContent and parsing it and making sure the original builder properties are preserved).

cbuescher committed:
Add fromXContent method to HighlightBuilder

For the search refactoring the HighlightBuilder needs a way to
create new instances by parsing xContent. For bwc this PR start
by moving over and slightly modifying the parsing from
HighlighterParseElement and keeps parsing for top level highlighter
and field options separate. Also adding tests for roundtrip
of random builder (rendering it to xContent and parsing it and
making sure the original builder properties are preserved)

cbuescher commented:
There's some weirdness going on with the first two commits I see, the diff looks okay though. Will try to squash this somehow.

colings86 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlightBuilder.java:
```diff
@@ -170,6 +172,114 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+            if (token == XContentParser.Token.FIELD_NAME) {
+                topLevelFieldName = parser.currentName();
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                if ("pre_tags".equals(topLevelFieldName) || "preTags".equals(topLevelFieldName)) {
```
Change we change the parsing to use ParseField? we have moved over the queries and are moving the aggs over to ParseField whilst refactorign so it would be good to include this in the rest of the search request refactoring too

cbuescher replied:
Can do, I mostly didn't do it so far because it's potentially error-prone and I wanted to minimize the diff). Would you be okay if we add ParseFields in HighlighterBuilder and leave the code as-is in HighlightParseElement (since this will go eventually)? That way we can also check in tests that we don't change anything during the refactoring.

colings86 replied:
Yes, that was actually what I meant (add ParseField to this class and leave HighlightParseElement as is)

colings86 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlightBuilder.java:
```diff
@@ -170,6 +172,114 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
+                            throw new IllegalArgumentException("If highlighter fields is an array it must contain objects containing a single field");
+                        }
+                    }
+                }
```
Should we add `else` blocks to catch illegal parameter names rather than just ignoring them?

cbuescher replied:
I'll check if this is easily possible here and add tests to it if possible.

colings86 replied:
In most other parsers (e.g. GeoBoundsParser) we do this by adding the following `else` block to the relevant places in the parser:

```java
} else if (!token(aggregationName, currentFieldName, token, parser, context.parseFieldMatcher(), otherOptions)) {
    throw new SearchParseException(context, "Unexpected token " + token + " [" + currentFieldName + "] in [" + aggregationName + "].", parser.getTokenLocation());
}
```

cbuescher replied:
@colings86 In most QueryParsers we seem to use ParsingException(XContentLocation contentLocation, String msg, Object... args), wonder what is the difference in case you know. Will use SearchParseException() for now though.

colings86 replied:
Sorry you are right, we chouls be using ParsingException. That snippet was the pre-refactored version. The difference is that ParsingException does not need the SearchContext (not available on the coordinating node) and actually points to the location in the request for the error (the XContentLocation). Please use ParsingException in this PR since this is going to be parsed on the coordinating node

cbuescher replied:
Yes, I just figured I don't have the search context but a parse context available, will go with Parsing Exception.

colings86 replied:
Thanks, it is probably also worth changing the other exceptions thrown in this method to ParsingException as it is a more appropriate exception to throw here compared with e.g. IllegalArguementException

colings86 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlighterParseElement.java:
```diff
@@ -94,7 +94,7 @@ public void parse(XContentParser parser, SearchContext context) throws Exception
     }
 
-    public SearchContextHighlight parse(XContentParser parser, QueryShardContext queryShardContext) throws IOException {
+    public static SearchContextHighlight parse(XContentParser parser, QueryShardContext queryShardContext) throws IOException {
```
why did we need to make these methods static? I think I have missed where this is used

cbuescher replied:
This doesn't have to be static, but since it doesn't need any object state I made it static to be able to call it from the new tests. Can change that back and create instance to call it on if you prefer, I'd be interested in your thoughts about pros/cons though.

colings86 commented:
@cbuescher This is looking good so far, I left a few comments

colings86 commented on core/src/test/java/org/elasticsearch/search/highlight/HighlightBuilderTests.java:
```diff
@@ -108,6 +126,83 @@ public void testEqualsAndHashcode() throws IOException {
+
+        context.reset(parser);
+        HighlightBuilder highlightBuilder = HighlightBuilder.fromXContent(context);
+        assertArrayEquals("setting tags_schema 'styled' should alter pre_tags", HighlighterParseElement.STYLED_PRE_TAG,
```
Can we move these default constants to HighlighterBuilder since HighlighterParseElement will be deleted?

cbuescher replied:
Makes sense, and is already part of a follow-up PR that's based on this, can also do this now but would prefer to leave it like is for now.

colings86 replied:
Makes sense if its in a future PR but we should make sure they are moved

cbuescher committed:
Switching HighlighterBuilder to use ParseFields

cbuescher committed:
Adding checks and tests for exceptions on unknown fieldnames

cbuescher commented:
@colings86 thanks for the review, I added commits that switch to using ParseFields in HighlightBuilder and added checks for unknown field names in the parser code (only in HighlightBuilder since the HighlightParseElement will be removed at some point in the future as far as I understand). Would be glad if you could take a second look.

colings86 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlightBuilder.java:
```diff
@@ -164,24 +169,138 @@ public Boolean useExplicitFieldOrder() {
+     *            call
+     * @return the new {@link HighlightBuilder}
+     */
+    public static HighlightBuilder fromXContent(QueryParseContext parseContext) throws IOException {
```
Can we not make the AbstractHighlighterBuilder define a fromXContent() method so it has the code for parsing the common attributes and then defer to a doFromXContent() to read the implementation specific attributes in HighlighterBuilder and Field? I am worried that if we duplicate this code in each implementation we will end up with parsing inconsistencies over time

cbuescher replied:
I had a look at this already and decided to not do it with this refactoring since the only solutions I found would change the parsing logic quiet a bit and my goal for this was to move over the parsing code from HighlightParseElement with as few modifications as possible. I agree that we should look into this, maybe in a follow-up PR though?

colings86 replied:
I think this should be done as part of the highlighter refactoring as it will make the code more robust against inconsistencies. Whether it's in this PR of a subsequent one I will leave to you but I think it should be done soon

cbuescher replied:
I opened https://github.com/elastic/elasticsearch/issues/15285 to track this.

cbuescher commented:
Hi @colings86, anything else to add here after ParseFields and additional exceptions?

colings86 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlightBuilder.java:
```diff
@@ -164,24 +169,138 @@ public Boolean useExplicitFieldOrder() {
+                } else {
+                    throw new ParsingException(parser.getTokenLocation(), "cannot parse object with name [{}]", topLevelFieldName);
+                }
+            }
```
We should add an else block here too so we throw an error if we get a token of a type we are not expecting

cbuescher replied:
Sure, will do.

colings86 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlightBuilder.java:
```diff
@@ -286,14 +405,88 @@ public void innerXContent(XContentBuilder builder) throws IOException {
+                    } else {
+                        throw new ParsingException(parser.getTokenLocation(), "cannot parse object with name [{}]", currentFieldName);
+                    }
+                }
```
We should add an else block here too so we throw an error if we get a token of a type we are not expecting

colings86 commented on core/src/test/java/org/elasticsearch/search/highlight/HighlightBuilderTests.java:
```diff
@@ -45,23 +62,26 @@
+        injectedQueryParsers.add(new MatchAllQueryParser());
+        injectedQueryParsers.add(new IdsQueryParser());
+        injectedQueryParsers.add(new TermQueryParser());
+        indicesQueriesRegistry = new IndicesQueriesRegistry(Settings.settingsBuilder().build(), injectedQueryParsers, namedWriteableRegistry);
```
Can this not use the injector approach we have used in the AbstractQueryTestCase and the SearchSourceBuilderTests? That way we won't have the hassle of registering each query builder/parser when we add a test that uses a new one.

cbuescher replied:
I think we shouldn't test with different query parsers here, since it complicates the setup and doesn't add anything to the test. We test serialization etc... of query parsers elsewhere and I tried hard to avoid the Juice/Injector stuff in this test, so unless we really need it (in which case we can add it later) I'd prefer to not use it here.

colings86 commented:
Left a couple of comments but otherwise LGTM

cbuescher committed:
Adding checks for unexpected tokens in parser

cbuescher commented:
Thanks for the review, added the checks for unexpected token while parsing, will merge in a bit.
